File: backend/app/api/routers/ai_platform.py
"""AI 平台管理路由（API.md §12，M8）。保持薄：解析请求、调用服务。

按权限码授权：admin.prompts / admin.models / admin.evaluations / admin.plugins。
"""

import uuid

from fastapi import APIRouter, Depends, Query

from app.api.dependencies import DbSession, require_permission
from app.models.user import User
from app.schemas.admin import DeleteResponse
from app.schemas.ai_platform import (
    EvaluationDetailResponse,
    EvaluationResultItem,
    EvaluationRunRequest,
    ModelConfigCreateRequest,
    ModelConfigItem,
    ModelConfigListResponse,
    ModelConfigUpdateRequest,
    PluginItem,
    PluginListResponse,
    PluginUpdateRequest,
    PromptActivateResponse,
    PromptVersionCreateRequest,
    PromptVersionItem,
    PromptVersionListResponse,
)
from app.schemas.common import Page
from app.services.evaluation_service import EvaluationService
from app.services.model_config_service import ModelConfigService
from app.services.plugin_service import PluginService
from app.services.prompt_service import PromptService

router = APIRouter(prefix="/admin", tags=["ai-platform"])

PromptsAdmin = Depends(require_permission("admin.prompts"))
ModelsAdmin = Depends(require_permission("admin.models"))
EvaluationsAdmin = Depends(require_permission("admin.evaluations"))
PluginsAdmin = Depends(require_permission("admin.plugins"))


def _to_prompt_item(row) -> PromptVersionItem:
    return PromptVersionItem(
        id=row.id,
        key=row.key,
        name=row.name,
        description=row.description,
        content=row.content,
        version=row.version,
        is_active=row.is_active,
        created_at=row.created_at,
    )


def _to_model_item(row) -> ModelConfigItem:
    return ModelConfigItem(
        id=row.id,
        name=row.name,
        kind=row.kind,
        provider=row.provider,
        model_name=row.model_name,
        base_url=row.base_url,
        api_key_ref=row.api_key_ref,
        is_active=row.is_active,
        priority=row.priority,
    )


def _to_evaluation_item(row) -> EvaluationResultItem:
    return EvaluationResultItem(
        id=row.id,
        name=row.name,
        status=row.status,
        total_questions=row.total_questions,
        passed=row.passed,
        created_at=row.created_at,
    )


def _to_plugin_item(row) -> PluginItem:
    return PluginItem(
        id=row.id,
        name=row.name,
        version=row.version,
        description=row.description,
        entry_point=row.entry_point,
        enabled=row.enabled,
    )


# ---------- Prompt 管理（§12.1） ----------


@router.get("/prompts", response_model=PromptVersionListResponse)
def list_prompts(
    session: DbSession, current_user: User = PromptsAdmin
) -> PromptVersionListResponse:
    rows = PromptService(session).list_versions()
    return PromptVersionListResponse(items=[_to_prompt_item(r) for r in rows])


@router.post("/prompts/{key}/versions", response_model=PromptVersionItem)
def create_prompt_version(
    key: str,
    payload: PromptVersionCreateRequest,
    session: DbSession,
    current_user: User = PromptsAdmin,
) -> PromptVersionItem:
    row = PromptService(session).create_version(
        key,
        content=payload.content,
        name=payload.name,
        description=payload.description,
        created_by=current_user.id,
    )
    return _to_prompt_item(row)


@router.post("/prompts/{prompt_id}/activate", response_model=PromptActivateResponse)
def activate_prompt(
    prompt_id: uuid.UUID,
    session: DbSession,
    current_user: User = PromptsAdmin,
) -> PromptActivateResponse:
    row = PromptService(session).activate(prompt_id)
    return PromptActivateResponse(id=row.id, is_active=row.is_active)


# ---------- 模型管理（§12.2） ----------


@router.get("/models", response_model=ModelConfigListResponse)
def list_models(
    session: DbSession, current_user: User = ModelsAdmin
) -> ModelConfigListResponse:
    rows = ModelConfigService(session).list()
    return ModelConfigListResponse(items=[_to_model_item(r) for r in rows])


@router.post("/models", response_model=ModelConfigItem)
def create_model(
    payload: ModelConfigCreateRequest,
    session: DbSession,
    current_user: User = ModelsAdmin,
) -> ModelConfigItem:
    return _to_model_item(ModelConfigService(session).create(payload))


@router.put("/models/{config_id}", response_model=ModelConfigItem)
def update_model(
    config_id: uuid.UUID,
    payload: ModelConfigUpdateRequest,
    session: DbSession,
    current_user: User = ModelsAdmin,
) -> ModelConfigItem:
    return _to_model_item(ModelConfigService(session).update(config_id, payload))


@router.delete("/models/{config_id}", response_model=DeleteResponse)
def delete_model(
    config_id: uuid.UUID,
    session: DbSession,
    current_user: User = ModelsAdmin,
) -> DeleteResponse:
    row = ModelConfigService(session).delete(config_id)
    return DeleteResponse(id=row.id, deleted=True)


# ---------- 评估（§12.3） ----------


@router.post("/evaluations", response_model=EvaluationDetailResponse)
def run_evaluation(
    payload: EvaluationRunRequest,
    session: DbSession,
    current_user: User = EvaluationsAdmin,
) -> EvaluationDetailResponse:
    row = EvaluationService(session).run(payload.name, payload.questions, current_user.id)
    return EvaluationDetailResponse(
        **_to_evaluation_item(row).model_dump(), details=row.details
    )


@router.get("/evaluations", response_model=Page[EvaluationResultItem])
def list_evaluations(
    session: DbSession,
    current_user: User = EvaluationsAdmin,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> Page[EvaluationResultItem]:
    rows, total = EvaluationService(session).list(page, page_size)
    return Page(
        items=[_to_evaluation_item(r) for r in rows],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get("/evaluations/{evaluation_id}", response_model=EvaluationDetailResponse)
def get_evaluation(
    evaluation_id: uuid.UUID,
    session: DbSession,
    current_user: User = EvaluationsAdmin,
) -> EvaluationDetailResponse:
    row = EvaluationService(session).get(evaluation_id)
    return EvaluationDetailResponse(
        **_to_evaluation_item(row).model_dump(), details=row.details
    )


# ---------- 插件（§12.4） ----------


@router.get("/plugins", response_model=PluginListResponse)
def list_plugins(
    session: DbSession, current_user: User = PluginsAdmin
) -> PluginListResponse:
    rows = PluginService(session).list()
    return PluginListResponse(items=[_to_plugin_item(r) for r in rows])


@router.put("/plugins/{plugin_id}", response_model=PluginItem)
def update_plugin(
    plugin_id: uuid.UUID,
    payload: PluginUpdateRequest,
    session: DbSession,
    current_user: User = PluginsAdmin,
) -> PluginItem:
    return _to_plugin_item(PluginService(session).set_enabled(plugin_id, payload.enabled))

File: frontend/src/lib/services/ai-platform.ts
import { api } from "../api-client";
import type { Paginated } from "./common";

/**
 * AI 平台管理服务(Milestone 8):严格对齐 docs/API.md §12(AI 平台管理)。
 * 全部端点位于 /api/admin 前缀下,仅 admin 角色可访问;后端校验为权威,
 * 前端只负责展示后端返回的可读错误。
 */

/** Prompt 版本对象。GET /api/admin/prompts 返回全部版本(含历史版本)。 */
export type PromptVersion = {
  id: string;
  key: string;
  name: string | null;
  description: string | null;
  content: string;
  version: number;
  is_active: boolean;
  created_at: string;
};

/** 创建 Prompt 新版本请求体(版本号由后端递增)。 */
export type PromptVersionCreateBody = {
  content: string;
  name?: string;
  description?: string;
};

/** 模型能力类型(定义权:docs/DATABASE.md model_configurations)。 */
export type ModelKind = "llm" | "vision" | "ocr" | "speech" | "embedding" | "reranker";

export const MODEL_KINDS: ModelKind[] = ["llm", "vision", "ocr", "speech", "embedding", "reranker"];

/** 模型配置对象。api_key_ref 只是密钥环境变量名,后端不返回密钥本身。 */
export type ModelConfiguration = {
  id: string;
  name: string;
  kind: ModelKind;
  provider: string;
  model_name: string;
  base_url: string | null;
  api_key_ref: string | null;
  is_active: boolean;
  priority: number;
};

/** 新建模型配置请求体。 */
export type ModelCreateBody = {
  name: string;
  kind: ModelKind;
  provider: string;
  model_name: string;
  base_url?: string;
  api_key_ref?: string;
  is_active?: boolean;
  priority?: number;
};

/** 更新模型配置请求体(字段可选,仅提交变更项)。 */
export type ModelUpdateBody = Partial<ModelCreateBody>;

/** 删除模型配置响应信封。 */
export type ModelDeleteResponse = {
  id: string;
  deleted: boolean;
};

/** 评估问题定义。expected_keywords 为空数组时不提交该字段。 */
export type EvaluationQuestion = {
  question: string;
  expected_keywords?: string[];
  require_source?: boolean;
};

/** 运行评估请求体(同步执行,真实调用 RAG+LLM 管线,可能耗时较长)。 */
export type EvaluationRunBody = {
  name: string;
  questions: EvaluationQuestion[];
};

/** 评估运行列表项。 */
export type EvaluationRun = {
  id: string;
  name: string;
  status: string;
  total_questions: number;
  passed: number;
  created_at: string;
};

/**
 * 评估运行详情:在列表项基础上附加逐题明细。
 * details 内部结构由后端定义,前端容错渲染(结构未知时原样展示 JSON)。
 */
export type EvaluationRunDetail = EvaluationRun & {
  details: unknown;
};

export type EvaluationListParams = {
  page?: number;
  page_size?: number;
};

/** 服务端插件注册项(定义权:docs/DATABASE.md plugins)。 */
export type AdminPlugin = {
  id: string;
  name: string;
  version: string | null;
  description: string | null;
  entry_point: string;
  enabled: boolean;
};

const base = "/api/admin";

export const aiPlatformService = {
  // ---- Prompt 管理 ----
  /** 返回全部 Prompt 的全部版本;前端按 key 分组展示。 */
  listPrompts: (signal?: AbortSignal) =>
    api.get<{ items: PromptVersion[] }>(`${base}/prompts`, { signal }),
  /** 基于给定内容创建某 key 的新版本(不自动生效)。 */
  createPromptVersion: (key: string, body: PromptVersionCreateBody) =>
    api.post<PromptVersion>(`${base}/prompts/${encodeURIComponent(key)}/versions`, body),
  /** 激活指定版本;后端保证同 key 仅一个生效版本。 */
  activatePrompt: (id: string) =>
    api.post<Pick<PromptVersion, "id" | "is_active">>(
      `${base}/prompts/${encodeURIComponent(id)}/activate`,
    ),

  // ---- 模型配置 ----
  listModels: (signal?: AbortSignal) =>
    api.get<{ items: ModelConfiguration[] }>(`${base}/models`, { signal }),
  createModel: (body: ModelCreateBody) => api.post<ModelConfiguration>(`${base}/models`, body),
  updateModel: (id: string, body: ModelUpdateBody) =>
    api.put<ModelConfiguration>(`${base}/models/${encodeURIComponent(id)}`, body),
  deleteModel: (id: string) =>
    api.delete<ModelDeleteResponse>(`${base}/models/${encodeURIComponent(id)}`),

  // ---- 评估运行 ----
  /** 同步执行评估:真实调用检索与模型管线,响应可能较慢。 */
  runEvaluation: (body: EvaluationRunBody) =>
    api.post<EvaluationRunDetail>(`${base}/evaluations`, body),
  listEvaluations: (params: EvaluationListParams = {}, signal?: AbortSignal) =>
    api.get<Paginated<EvaluationRun>>(`${base}/evaluations`, {
      query: { page: params.page, page_size: params.page_size },
      signal,
    }),
  getEvaluation: (id: string, signal?: AbortSignal) =>
    api.get<EvaluationRunDetail>(`${base}/evaluations/${encodeURIComponent(id)}`, { signal }),

  // ---- 插件管理 ----
  listPlugins: (signal?: AbortSignal) =>
    api.get<{ items: AdminPlugin[] }>(`${base}/plugins`, { signal }),
  /** 仅支持切换启用状态;失败时前端回滚开关并展示后端可读错误。 */
  updatePlugin: (id: string, enabled: boolean) =>
    api.put<AdminPlugin>(`${base}/plugins/${encodeURIComponent(id)}`, { enabled }),
};
